### PR TITLE
make route stats context as pure class

### PR DESCRIPTION
**Pull request by ramaraochavali**
Commit Message:make route stats context as pure class
Additional Description:make route stats context as pure class
Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@jmarantz I realized there is already a `context.h` in router package that we can use. Changed it like that. PTAL

**Review by jmarantz — COMMENTED** [review]
nice; thank you!

one tiny comment otherwise lgtm.

coverage failure looks like a flake; it might pass when you remove the 'explicit'.

**Review by jmarantz — DISMISSED** [review]

File: envoy/router/context.h [hunk 1]
```diff
@@ -35,5 +35,23 @@ class Context {
...
+ * RouteStatsContext defines config needed to generate all route level stats.
+ */
+class RouteStatsContext {
+public:
+  virtual ~RouteStatsContext() = default;
+
+  /**
+   * @return stat name for route specific stats.
+   */
+  virtual const Stats::StatName statName() const PURE;
```
Comment (htuch) [inline envoy/router/context.h:48]: What does it mean to return a `const` of a value?

Comment (ramaraochavali) [reply envoy/router/context.h:48]: yeah. it is not needed. Will remove

**Review by htuch — COMMENTED** [review]
LGTM, but I'm curious on why this change? Is there some testing or multi-implementation use case? Just asking as on its own, it's not an ideal to make everything fully abstract, e.g. implementation local classes can be just that.

**Comment by ramaraochavali** [comment]
@htuch sorry for not providing context on this in PR. @jmarantz suggested this in another PR https://github.com/envoyproxy/envoy/pull/21302#discussion_r895996089 - but did not do it in that PR.

**Comment by jmarantz** [comment]
Yeah I suggested it because this PR eliminates the need to have a somewhat complex constructor inlined into a interface file.

**Review by htuch — APPROVED** [review]
Great, thanks!

**Pull request merged**